File: kotlin/src/main/kotlin/problem15/3sum.kt
package problem15

class Solution {
    fun threeSum(nums: IntArray): List<List<Int>> {
        nums.sort()
        val res = mutableListOf<List<Int>>()
        for (i in 0..(nums.size - 3)) {
            if (nums[i] > 0) {
                return res
            }
            if (i > 0 && nums[i - 1] == nums[i]) {
                continue
            }
            for (j in (i + 1)..(nums.size - 2)) {
                if (j > (i + 1) && nums[j - 1] == nums[j]) {
                    continue
                }
                val k = nums.binarySearch(-nums[i] - nums[j], j + 1)
                if (k > 0) {
                    res.add(listOf(nums[i], nums[j], nums[k]))
                }
            }
        }
        return res
    }
}

File: go/explore/october2021/day28/3sum.go
package main

import (
	"fmt"
	"sort"
)

func main() {
	fmt.Println(threeSum([]int{-1, 0, 1, 2, -1, -4}))
}

func threeSum(nums []int) [][]int {
	n := len(nums)
	sort.Ints(nums)

	res := [][]int{}
	for i := 0; i < n-2; i++ {
		if i > 0 && nums[i-1] == nums[i] {
			continue
		}
		r := n - 1
		for l := i + 1; l < r; l++ {
			if l > i+1 && nums[l-1] == nums[l] {
				continue
			}
			for l+1 < r && nums[i]+nums[l]+nums[r] > 0 {
				r--
			}
			if nums[i]+nums[l]+nums[r] == 0 {
				res = append(res, []int{nums[i], nums[l], nums[r]})
			}
		}
	}
	return res
}
